pkg/venafi/tpp/connector.go

```diff
@@ -1232,25 +1244,65 @@ func (c *Connector) RetrieveCertificate(req *certificate.Request) (certificates
 	}
 
 	startTime := time.Now()
+	retrieveCount := 0
 	for {
-		var retrieveResponse *certificateRetrieveResponse
-		retrieveResponse, err = c.retrieveCertificateOnce(certReq)
+		retrieveCount++
+		statusCode, _, body, err := c.request("POST", urlResourceCertificateRetrieve, certReq)
 		if err != nil {
 			return nil, fmt.Errorf("unable to retrieve: %s", err)
 		}
-		if retrieveResponse.CertificateData != "" {
-			certificates, err = newPEMCollectionFromResponse(retrieveResponse.CertificateData, req.ChainOption)
+		var resp certificateRetrieveResponse
+		err = json.Unmarshal(body, &resp)
+		if err != nil {
+			return nil, fmt.Errorf("retrieve response body is unexpected: %s, HTTP status was %d and body was: %s", err, statusCode, body)
+		}
+		switch {
+		// After many tests, we found that the only way to know if the current
+		// certificate request is stuck due to an old failed enrollment is to
+		// look for "WebSDK CertRequest" or "Fix any errors, and then click
+		// Retry." in the retrieve response. We only check this on the first
+		// retrieve because we know that these two messages always disappear
+		// after resetting.
+		case statusCode == http.StatusInternalServerError && retrieveCount == 1 &&
+			(resp.Status == "This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry." ||
+				resp.Status == "WebSDK CertRequest Module Requested Certificate"):
+			statusCode, status, body, err := c.request("POST", urlResourceCertificateReset, certificateResetRequest{
+				CertificateDN: req.PickupID,
+				Restart:       true,
+			})
+			if err != nil {
+				return nil, fmt.Errorf("while resetting certificate due to a past failed enrollment: %w", err)
+			}
+			if statusCode != http.StatusOK {
+				return nil, fmt.Errorf("while resetting certificate due to a past failed enrollment: %s, body: '%s'", status, body)
+			}
+
+			continue
+		case statusCode == http.StatusInternalServerError && resp.Status != "":
+			return nil, fmt.Errorf("enrollment failed at stage %d: %s", resp.Stage, resp.Status)
+		case statusCode == http.StatusOK:
```

Comment:
Instead of checking for `http.StatusOK` or `http.StatusAccepted`, can we just check `resp.CertificateData != ""` identical to the original code?

Reply:
I can replace `http.StatusOK` with `resp.CertificateData != ""`, but I would not be able to also do that for `http.StatusAccepted`, since I would fall into the `default` case (which results in an error).

Reply:
Looking at the [`POST /retrieve` documentation](https://docs.venafi.com/Docs/current/TopNav/Content/SDK/WebSDK/r-SDK-POST-Certificates-retrieve.php), it seems like 200 is the only HTTP code for which `CertificateData` will be present. For 202, the only documented JSON fields are `Status` and `Stage`, `CertificateData` isn't meant to be returned on a 202.

But I'd prefer someone confirming that 202 never returns a `CertificateData`.